Use the release branch of CoreCLR

tonybaloney opened the pull request:
There is a bug in the referenced CoreCLR submodule in master. 

The NuGet config in CoreCLR/tests/src/runtime (tests) references a feed on MyGet that no longer exists.
https://github.com/dotnet/coreclr/blob/release/1.0.0-rc2/tests/src/NuGet.Config

When you run BuildDeps.bat it will try and build CoreCLR which fails at the restore packages phase, since the feed no longer exists.

tonybaloney committed:
Use the release branch of CoreCLR

msftclas commented:
Hi __@tonybaloney__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        This seems like a small (but important) contribution, so no Contribution License Agreement is required at this point. Real humans will now evaluate your PR.
    </span>

TTYL, MSBOT;

tonybaloney commented:
`git pull origin release/1.0.0-rc2` from within the CoreCLR project will fix this for now.

tonybaloney commented:
@brettcannon I got stuck because the patch files are no longer correct. I'm rebuilding the patches like this:

```
cd CoreCLR\
git reset --hard
git checkout b16ff5935ff9df3211798f18f0c951666ae27774
cd ..
PatchDeps.bat
cd CoreCLR\
git fetch -t
git merge v1.0.0-rc2
git stash save --keep-index
git merge v1.0.0-rc2
git stash apply
```

In future I guess the patches would be merged into the CoreCLR and Python repositories respectively? It would be far easier to maintain forks of CoreCLR and Python with a branch containing the commits.

tonybaloney committed:
Update patches to CoreCLR for rc2

tonybaloney commented:
@brettcannon @DinoV  It wasn't quite as simple as rebasing, I've updated the patches to reflect what was done, but the build.cmd in particular has changed, so a simple merge wouldn't work .

https://github.com/Microsoft/Pyjion/pull/187/commits/7cc141066346148588cd8e04b9e9efefb53a180e

tonybaloney commented:
```
c:\users\anthony\documents\github\pyjion\coreclr\src\vm\clrtocomcall.h(17): fatal error C1189: #error:  FEATURE_COMINTEROP is required for this file (compiling source file C:\Users\Anthony\Documents\github\Pyjion\CoreCLR\src\vm\classcom
```
working on this.

brettcannon commented:
Just to answer the "why not fork?" question: we didn't do that initially because we had contemplated talking to the CoreCLR team about taking our work upstream (they've said they would consider it). We have also contemplated trying out Chakra as a JIT back-end which makes us always on the fence about the work to fork CoreCLR itself as the patches need updates every few months, but not often enough to halt development.

tonybaloney commented:
Because the CoreCLR team refactored the build scripts for Rc2, the patches no longer apply, nor would a merge make sense. I had some merge conflicts and was honestly a little stumped about why the patches removed a bunch of projects from the build and then disables com interop. 
This PR in its current state still does not build because of the number of errors thrown at compile saying module depends on com interop feature. 
Is there a technical reason to why com interop needs to be disabled then maybe I can try and reformat the build scripts correctly. 
I guess the simplest option would be that the CoreCLR team manage a build of this that feature disabled, instead of compiling from source. 

The Python code works fine. 

In its current state Pyjion won't build because the Pinned version of CoreCLR references a series of nuget feeds that no longer exist. To use the new feeds you also need to bring in the rest of the commit history which includes the issues I shared above. 

I'm guessing both @brettcannon and @DinoV have cached binaries so this isn't an issue, but for anyone wanting to play, this is blocking the project from being usable. 

Seems a shame seeming as you've put all the work in but as CoreCLR is still a moving target it can't be avoided.

tonybaloney commented:
Awesome work so far tho, I spent a while reading the source to understand how it worked, really cool idea.

tonybaloney committed:
update patches

tintoy committed:
Merge pull request #1 from tonybaloney/update_coreclr

Update patches to fix breakages for CoreCLR 1.0.0-rc2.

tintoy committed:
Use release/1.0.0-rc2 branch of CoreCLR, and fix resulting breakages. Also, don't build mscorlib because it's not needed for Pyjion

tonybaloney committed:
Merge pull request #1 from tintoy/update_coreclr

Use release branch of CoreCLR and fix resulting breakages.

tonybaloney commented:
My more talented colleague @tintoy figured it out. this is all up to date now.

tonybaloney committed:
Extended implementation of ICorJitInfo with new methods added in CoreCLR project

tintoy committed:
Remove reference to mscoree from longfilepathwrappers (unused by Pyjion).

tonybaloney committed:
Merge pull request #2 from tintoy/update_coreclr

Remove reference to mscoree from longfilepathwrappers

tonybaloney commented:
PR completed. please review @brettcannon @DinoV @tintoy 

changes
- Updated patches to support RC2
- Change the CoreCLR build to skip mscorlib
- Removed references to the APPX feature
- Updated build for CoreCLR to use feature flag after CoreCLR team refactored `build.cmd`
- Removed references to mscoree from https://github.com/tonybaloney/Pyjion/blob/38944622de1377866b1bd591e60eb03e9c7ec8e6/Patches/CoreCLR/src/utilcode/longfilepathwrappers.cpp in a patchfile
- Fixed build error for pycomp.cpp, I commented out the reference to a function which does nothing, the method signature has changed. Not sure if this is correct. https://github.com/tonybaloney/Pyjion/blob/38944622de1377866b1bd591e60eb03e9c7ec8e6/Pyjion/pycomp.cpp
- Updated jitinfo.h, in which CorJitInfo implements the interface ICorJitInfo, which has additional abstracts in rc2 that did not exist previously https://github.com/tonybaloney/Pyjion/blob/38944622de1377866b1bd591e60eb03e9c7ec8e6/Pyjion/jitinfo.h

msftclas commented:
__@tonybaloney__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

brettcannon commented:
With PyCon less than a week away, @tonybaloney , we probably won't get to reviewing this until after we are back from the conference (but there's always a chance we will get to this before then). But thanks ahead of time for putting in all of this work to fix this!

And to answer some of you questions:
1. Yes, we both have previously built DLLs for CoreCLR, hence why we have not noticed the breakage (and why we need to get CI set up).
2. The COM interop stuff was turned off to simply avoid the overhead; we don't need it and @DinoV thought it would be better perf-wise to leave it all out.

DinoV commented on Pyjion/pycomp.cpp:
```diff
@@ -81,7 +81,7 @@ void CeeInit() {
     cccallbacks.m_hmodCoreCLR = (HINSTANCE)GetModuleHandleW(NULL);
     cccallbacks.m_pfnIEE = IEE;
-    cccallbacks.m_pfnGetCORSystemDirectory = GetCORSystemDirectoryInternal;
+    //cccallbacks.m_pfnGetCORSystemDirectory = GetCORSystemDirectoryInternal;
```
Instead of commenting this out can you update GetCORSystemDirectoryInternal to take a SString& ?  I realize we don't expect this to be called right now, but it'd be nice if it were obvious if it started getting called in the future.

tonybaloney replied:
OK done.

DinoV commented:
FYI the reason why we disabled COM interop was just because it was causing crashes when we would go to run our tests.  I never did track down what was the underlying cause, but it does appear that issue is gone.

While all of our tests pass with this change it appears that running the CPython test suite is failing.  When running test_argparse under windbg (e.g. PyjionPull\Python\PCbuild\amd64>C:\debuggers\windbg.exe .\python_d.exe ..\..\Lib\test\test_argparse.py) I see that we're blowing up because the JIT is asserting that the default page size is zero.  It appears that the JIT now needs a call to jitStartup with a ICorJitHost* instance.  

It's a pretty simple interface to implement, would you be willing to modify cee.h so CExecutionEngine implements this interface as well and that we call jitStartup with it in pycomp.cpp CeeInit?

Hopefully that will get everything running again and then I think the change should be good to merge!

tonybaloney committed:
Uncomment GetCORSystemDirectoryInternal assignment and update method signature to match new data struct

tonybaloney committed:
Fix issue where test_argparse is expecting jitStartup to be called

tonybaloney commented:
OK @DinoV I fixed that issue, the tests are failing (it seems) right at the end, cleaning up the registers. I assume that where I instantiated the g_jitHost global was wrong. I'm not really clear about the execution order here, we have a series of globals and a lot of module code declared in headers. Could you give me a pointer?

DinoV commented on Pyjion/cee.h:
```diff
@@ -358,6 +359,31 @@ class CExecutionEngine : public IExecutionEngine, public IEEMemoryManager {
+
+	int getIntConfigValue(const wchar_t * name, int defaultValue)
+	{
+		return 0;
```
This should return defaultValue

DinoV commented:
It looks like jitInfo.h getRelocTypeHint is being hit now too, can you remove the printf in there as well?

I think g_jitHost should basically never go away (it'll get cleaned up when pyjit.dll gets unloaded, which practically may never really happen).

I'm running the tests now, I'll see if I get the same failure and let you know what I figure out about it.

DinoV commented:
I didn't hit the crash when running the tests, so everything looks good now other than getIntConfigValue not returning default value.  If you can fix that up I'll merge the change.  Thanks, this is a great update!

tonybaloney committed:
set getIntConfigValue to ret default Value instead of 0, removed unnessecary printf

tonybaloney commented:
According to the docs:
```
// ICorJitHost
//
// ICorJitHost provides the interface that the JIT uses to access some functionality that
// would normally be provided by the operating system. This is intended to allow for
// host-specific policies re: memory allocation, configuration value access, etc. It is
// expected that the `ICorJitHost` value provided to `jitStartup` lives at least as
// long as the JIT itself.
```
So I think I instantiated g_jitHost in the correct place.

Fixed the other 2 requests. @DinoV  all good to go

DinoV commented:
Thanks a lot for tackling this!
